[#63594] Primerize Project copy form

myabc opened the pull request:
⚠️ **~~This PR is based on #18716, #18478. Please review/merge those PRs first.~~**

# Ticket

https://community.openproject.org/wp/63594

# What are you trying to accomplish?

Updates the form visuals based on new mockups: the Project copy form now only includes required (built-in and custom) fields, along with the "Subproject of" field.

In detail:

* Introduces `Projects::CopyFormComponent`.
* Adds `Projects::CopyForm` with basic specs.
* Adds `ProjectsController#copy_form` action, permissions and route mappings.
* Updates copy project feature specs to reflect form changes.
* Removes the old Angular `CopyProjectComponent` and remainders of `OpenprojectProjectsModule`.
* Ensures both the create and copy forms correctly handle the scenario where no Copy Options dependencies or associations are selected.

## Screenshots

![Screenshot 2025-06-25 at 11 42 23](https://github.com/user-attachments/assets/341f03c7-e666-449b-b228-60ec6b49781e)
![Screenshot 2025-06-25 at 11 42 40](https://github.com/user-attachments/assets/cd274396-5914-4768-9085-547b7d558afc)
![Screenshot 2025-06-25 at 11 42 55](https://github.com/user-attachments/assets/88077ad7-7ea7-4277-9956-632ee3dea1f9)


# What approach did you choose and why?

This work is similar in approach to #18716.

N.B. This PR does not fully implement the design specified in the latest iteration of the mockups - follow up work will take place in [OP 65026](https://community.openproject.org/wp/65026).

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
[#63594] Primerize Project copy form

Updates the form visuals based on new mockups: the Project copy form now
only includes required (built-in and custom) fields, along with the
"Subproject of" field.

In detail:

* Introduces `Projects::CopyFormComponent`.
* Adds `Projects::CopyForm` with basic specs.
* Adds `ProjectsController#copy_form` action, permissions and route
  mappings.
* Updates copy project feature specs to reflect form changes.

myabc committed:
[ng] Remove CopyProjectComponent

myabc committed:
[ng] Remove remainder of OpenprojectProjectsModule

myabc committed:
Fix submission when no copy options are checked

Ensures no asssocations/dependencies can be checked when the user copies
or creates a project from template.

myabc committed:
Remove unused CopyOptions#dependencies validation

cbliard commented on spec/controllers/projects_controller_spec.rb:
```diff
@@ -305,25 +306,90 @@
-      expect(response).to be_successful
-      expect(response).to render_template "copy"
+    shared_examples_for "successful request" do
+      it "renders 'copy_form'", :aggregate_faulures do
```
typo: `:aggregate_faulures` => `aggregate_failures`

```suggestion
      it "renders 'copy_form'", :aggregate_failures do
```

myabc replied:
good catch! thanks

cbliard commented on spec/controllers/projects_controller_spec.rb:
```diff
@@ -305,25 +306,90 @@
+
+    context "as a non-admin with copy projects permission" do
+      let(:user) { create(:user, member_with_roles: { project => project_role }) }
+      let(:project_role) { build(:project_role, permissions: [:copy_projects]) }
```
This works, but just as a FYI, there is also a `member_with_permissions` parameter available which takes care of creating the `ProjectRole`. I find it more readable as there is less boilerplate code.

It would then look like this:

```ruby
let(:user) { create(:user, member_with_permissions: { project => [:copy_projects] }) }
```

cbliard commented on spec/features/projects/copy_spec.rb:
```diff
@@ -178,16 +178,11 @@
       end
 
       it "separates optional and required custom fields for new" do
```
```suggestion
      it "renders only required project attributes" do
```

cbliard reviewed (changes requested):
Nice 🍍 
I only found one issue:
- if there is an error at submitting the copy like for instance an empty project name, then an error is displayed next screen but the user menu becomes unusable: clicking on the avatar icon in the top-right corner does not display the user menu anymore. I can only reproduce the issue on the feature branch. It works correctly on `dev`.

This is the only reason I'm marking the PR as "Request changes".

For the rest, they are only minor remarks which can be ignored or addressed in a later PR.

Also `js.forms.advanced_settings` I18n key can be removed from `js-en.yml`, and for `js.project.copy.copy_options`, I think it should migrate to `en.yml` even if that means it will have to be retranslated.

myabc committed:
Fix typo in ProjectsController spec metadata

Co-Authored-By: Christophe Bliard <176055+cbliard@users.noreply.github.com>

myabc committed:
DRY up user creation in ProjectsController spec

Co-Authored-By: Christophe Bliard <176055+cbliard@users.noreply.github.com>

myabc commented:
@cbliard Thanks again for your feedback! I will have to address the issue with the top menu bar in a separate BUG WP/PR as I realise it also affects other project pages.

> For the rest, they are only minor remarks which can be ignored or addressed in a later PR.

I've addressed these remarks. 

> Also `js.forms.advanced_settings I18n key can be removed from `js-en.yml`, and for js.project.copy.copy_options`, I think it should migrate to `en.yml` even if that means it will have to be retranslated.

Good point - I've noticed that we (as a team) tend to avoid renaming translation strings, but it's good to have strings in the right place. (with translation memories and AI, retranslating shouldn't be a huge issue). I'll make a note to do this as part of #18715 ([OP 65026](https://community.openproject.org/wp/65026)).

cbliard reviewed (approved):
Thanks for incorporating my suggestions

cbliard commented:
Failing feature tests `modules/gantt/spec/features/timeline/timeline_dates_spec.rb:205` and `spec/features/work_packages/display_representations/switch_display_representations_on_mobile_spec.rb:60` are passing locally.
